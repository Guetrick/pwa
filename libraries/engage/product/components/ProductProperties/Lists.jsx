import React from 'react';
import PropTypes from 'prop-types';
import { getSubgroupsFromProperties } from './helpers/getSubgroupsFromProperties';
import { getPropertiesWithoutSubgroup } from './helpers/getPropertiesWithoutSubgroup';
import { getPropertiesBySubgroup } from './helpers/getPropertiesBySubgroup';
import Group from './Group';
import Rows from './Rows';

/**
 * Renders lists of properties that are eventually grouped.
 * @param {Object} props The component props.
 * @return {JSX}
 */
const Lists = ({ properties }) => (
  <React.Fragment>
    <Rows properties={getPropertiesWithoutSubgroup(properties)} />
    {getSubgroupsFromProperties(properties).map(group => (
      <React.Fragment key={group}>
        <Group group={group} />
        <Rows properties={getPropertiesBySubgroup(properties, group)} />
      </React.Fragment>
    ))}
  </React.Fragment>
);

Lists.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default Lists;
