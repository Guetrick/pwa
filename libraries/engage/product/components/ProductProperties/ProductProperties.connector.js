import { connect } from 'react-redux';
import { makeGetProductProperties } from '../../selectors/product';

/**
 * @return {Function}
 */
function makeMapStateToProps() {
  const getProductProperties = makeGetProductProperties();

  return (state, props) => ({
    properties: getProductProperties(state, props),
  });
}

export default connect(makeMapStateToProps);
