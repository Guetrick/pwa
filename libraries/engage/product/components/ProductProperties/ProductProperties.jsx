import React from 'react';
import PropTypes from 'prop-types';
import { isBeta } from '../../../core';
import { ProductPropertiesSimple } from './ProductPropertiesSimple';
import { ProductPropertiesGrouped } from './ProductPropertiesGrouped';
import connect from './ProductProperties.connector';

/**
 * The product properties.
 * @returns {JSX}
 */
function ProductPropertiesContainer({ properties }) {
  if (!properties) {
    return null;
  }

  if (!isBeta()) { // TODO: Can be removed once the beta phase (product) is over!
    return <ProductPropertiesSimple properties={properties} />;
  }

  /*
    This feature is currently in BETA testing.
    It should only be used for approved BETA Client Projects
  */
  return (
    <ProductPropertiesGrouped properties={properties} />
  );
}

ProductPropertiesContainer.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesContainer.defaultProps = {
  properties: null,
};

export const ProductProperties = connect(ProductPropertiesContainer);
