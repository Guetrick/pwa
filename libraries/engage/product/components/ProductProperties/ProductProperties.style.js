import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { variables: { gap } = {} } = themeConfig;

export const container = css({
  fontSize: '0.875rem',
  padding: `0 ${gap.big}px ${gap.big}px`,
  marginBottom: 12,
});

export const label = css({
  paddingBottom: (gap.small * 1.5),
});

export const tableCell = css({
  maxWidth: 100,
  padding: `${gap.xsmall * 0.5}px ${gap.small}px`,
  overflowWrap: 'break-word',
  ':first-of-type': {
    paddingLeft: 0,
  },
  ':last-of-type': {
    paddingRight: 0,
  },
});
