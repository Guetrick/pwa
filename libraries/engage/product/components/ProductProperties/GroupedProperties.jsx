import React from 'react';
import PropTypes from 'prop-types';
import { Accordion } from '@shopgate/pwa-ui-material';
import Lists from './Lists';
import Wrapper from './Wrapper';
import { accordion } from './style';

/**
 * Renders the properties as groups.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const GroupedProperties = ({ properties, groups }) => groups
  .map(group => (
    <div key={group} className={accordion}>
      <Accordion renderLabel={() => group} testId={`product-properties-group-${group}`}>
        <Wrapper dense>
          <Lists
            properties={properties.filter(property => property.displayGroup === group)}
          />
        </Wrapper>
      </Accordion>
    </div>
  ));

GroupedProperties.propTypes = {
  groups: PropTypes.arrayOf(PropTypes.string).isRequired,
  properties: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default GroupedProperties;
