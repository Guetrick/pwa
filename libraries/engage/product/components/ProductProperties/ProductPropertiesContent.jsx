import React from 'react';
import PropTypes from 'prop-types';
import { isBeta } from '../../../core';
import { getGroupsFromProperties } from './helpers/getGroupsFromProperties';
import ProductPropertiesGrouped from './ProductPropertiesGrouped';
import ProductPropertiesWrapper from './ProductPropertiesWrapper';
import ProductPropertiesRows from './ProductPropertiesRows';
import { groupsContainer } from './ProductProperties.style';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const ProductPropertiesContent = ({ properties }) => {
  if (!properties) {
    return null;
  }

  const groups = getGroupsFromProperties(properties);

  // Display the simple properties if no groups exist or if not in beta mode.
  if (!isBeta() || !groups || groups.length === 0) {
    return (
      <ProductPropertiesWrapper>
        <ProductPropertiesRows properties={properties} />
      </ProductPropertiesWrapper>
    );
  }

  /*
    This feature is currently in BETA testing.
    It should only be used for approved BETA Client Projects
  */
  return (
    <div className={groupsContainer}>
      <ProductPropertiesGrouped properties={properties} groups={groups} />
    </div>
  );
};

ProductPropertiesContent.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesContent.defaultProps = {
  properties: null,
};

export default ProductPropertiesContent;
