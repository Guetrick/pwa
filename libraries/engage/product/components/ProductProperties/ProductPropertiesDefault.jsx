import React from 'react';
import PropTypes from 'prop-types';
import { SurroundPortals } from '@shopgate/pwa-common/components';
import { PRODUCT_PROPERTIES } from '@shopgate/pwa-common-commerce/product/constants/Portals';
import { ProductPropertiesRows } from './ProductPropertiesRows';
import { container } from './ProductProperties.style';
import connect from './ProductPropertiesDefault.connector';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ProductPropertiesDefaultContainer({ properties }) {
  return (
    <SurroundPortals portalName={PRODUCT_PROPERTIES}>
      {(properties !== null && properties.length !== 0) && (
        <div className={container}>
          <table>
            <tbody>
              <ProductPropertiesRows properties={properties} />
            </tbody>
          </table>
        </div>
      )}
    </SurroundPortals>
  );
}

ProductPropertiesDefaultContainer.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesDefaultContainer.defaultProps = {
  properties: null,
};

export const ProductPropertiesDefault = connect(ProductPropertiesDefaultContainer);
