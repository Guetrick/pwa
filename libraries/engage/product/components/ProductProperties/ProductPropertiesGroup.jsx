import React from 'react';
import PropTypes from 'prop-types';
import { subgroup } from './ProductProperties.style';

/**
 * @param {Object} props The component props.
 * @return {JSX}
 */
const ProductPropertiesGroup = ({ group }) => (
  <tr>
    <td colSpan="2" className={subgroup}>{group}</td>
  </tr>
);

ProductPropertiesGroup.propTypes = {
  group: PropTypes.string.isRequired,
};

export default React.memo(ProductPropertiesGroup);
