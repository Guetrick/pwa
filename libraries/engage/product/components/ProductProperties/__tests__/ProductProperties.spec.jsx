import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import mockConsole from 'jest-mock-console';
import mockRenderOptions from '@shopgate/pwa-common/helpers/mocks/mockRenderOptions';
import ProductProperties from '../ProductProperties';
import { makeGetProductProperties } from '../../../selectors/product';

const mockStore = configureStore();
const store = mockStore({});

jest.mock('../../../selectors/product', () => ({
  makeGetProductProperties: jest.fn(() => jest.fn()),
}));

const properties = [
  { displayGroup: 'test' },
];

describe('<ProductProperties />', () => {
  let restoreConsole;

  beforeEach(() => {
    restoreConsole = mockConsole();
  });

  afterEach(() => {
    restoreConsole();
  });

  it('should not render if no properties where passed', () => {
    makeGetProductProperties.mockReturnValueOnce(() => null);
    const wrapper = mount(<ProductProperties key="1" store={store} />, mockRenderOptions);
    expect(wrapper.find('ProductProperties').instance()).toEqual(null);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render if properties are passed', () => {
    makeGetProductProperties.mockReturnValueOnce(() => properties);
    const wrapper = mount(<ProductProperties key="2" store={store} />, mockRenderOptions);
    expect(wrapper.find('Content').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
});
