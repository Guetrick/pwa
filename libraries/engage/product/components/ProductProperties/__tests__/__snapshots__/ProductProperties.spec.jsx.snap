// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductProperties /> should not render if no properties where passed 1`] = `
<Connect(ProductProperties)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ProductProperties
    dispatch={[Function]}
    properties={null}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <SurroundPortals
      portalName="product.properties"
      portalProps={
        Object {
          "properties": null,
        }
      }
    >
      <Portal
        name="product.properties.before"
        props={
          Object {
            "properties": null,
          }
        }
      />
      <Portal
        name="product.properties"
        props={
          Object {
            "properties": null,
          }
        }
      >
        <ProductPropertiesContent
          properties={null}
        />
      </Portal>
      <Portal
        name="product.properties.after"
        props={
          Object {
            "properties": null,
          }
        }
      />
    </SurroundPortals>
  </ProductProperties>
</Connect(ProductProperties)>
`;

exports[`<ProductProperties /> should render if properties are passed 1`] = `
<Connect(ProductProperties)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ProductProperties
    dispatch={[Function]}
    properties={
      Array [
        Object {
          "displayGroup": "test",
        },
      ]
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <SurroundPortals
      portalName="product.properties"
      portalProps={
        Object {
          "properties": Array [
            Object {
              "displayGroup": "test",
            },
          ],
        }
      }
    >
      <Portal
        name="product.properties.before"
        props={
          Object {
            "properties": Array [
              Object {
                "displayGroup": "test",
              },
            ],
          }
        }
      />
      <Portal
        name="product.properties"
        props={
          Object {
            "properties": Array [
              Object {
                "displayGroup": "test",
              },
            ],
          }
        }
      >
        <ProductPropertiesContent
          properties={
            Array [
              Object {
                "displayGroup": "test",
              },
            ]
          }
        >
          <ProductPropertiesWrapper>
            <div
              className="css-1x8awm9"
            >
              <table>
                <tbody>
                  <ProductPropertiesRows
                    properties={
                      Array [
                        Object {
                          "displayGroup": "test",
                        },
                      ]
                    }
                  >
                    <ProductPropertiesRow
                      key="undefined-undefined"
                    >
                      <tr
                        key="undefinedundefined"
                      >
                        <td
                          className={
                            Object {
                              "data-css-fl9ufm": "",
                            }
                          }
                        />
                        <td
                          className={
                            Object {
                              "data-css-fl9ufm": "",
                            }
                          }
                          data-test-id="property: undefined"
                        />
                      </tr>
                    </ProductPropertiesRow>
                  </ProductPropertiesRows>
                </tbody>
              </table>
            </div>
          </ProductPropertiesWrapper>
        </ProductPropertiesContent>
      </Portal>
      <Portal
        name="product.properties.after"
        props={
          Object {
            "properties": Array [
              Object {
                "displayGroup": "test",
              },
            ],
          }
        }
      />
    </SurroundPortals>
  </ProductProperties>
</Connect(ProductProperties)>
`;
